drivers/eks/eks_driver.go

```diff
@@ -927,3 +1002,148 @@ func (d *Driver) GetK8SCapabilities(ctx context.Context, _ *types.DriverOptions)
 		},
 	}, nil
 }
+
+func (d *Driver) GetVersion(ctx context.Context, info *types.ClusterInfo) (*types.KubernetesVersion, error) {
+	cluster, err := d.getClusterStats(ctx, info)
+	if err != nil {
+		return nil, err
+	}
+
+	version := &types.KubernetesVersion{Version: *cluster.Version}
+
+	return version, nil
+}
+
+func (d *Driver) getClusterStats(ctx context.Context, info *types.ClusterInfo) (*eks.Cluster, error) {
+	state, err := getState(info)
+	if err != nil {
+		return nil, err
+	}
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(state.Region),
+		Credentials: credentials.NewStaticCredentials(
+			state.ClientID,
+			state.ClientSecret,
+			state.SessionToken,
+		),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	svc := eks.New(sess)
+	cluster, err := svc.DescribeClusterWithContext(ctx, &eks.DescribeClusterInput{
+		Name: aws.String(state.DisplayName),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return cluster.Cluster, nil
+}
+
+func (d *Driver) SetVersion(ctx context.Context, info *types.ClusterInfo, version *types.KubernetesVersion) error {
+	logrus.Info("updating kubernetes version")
+	state, err := getState(info)
+	if err != nil {
+		return err
+	}
+
+	state.KubernetesVersion = version.Version
+	if err := d.updateClusterAndWait(ctx, state); err != nil {
+		return err
+	}
+
+	logrus.Info("kubernetes version update success")
+	return nil
+}
+
+func (d *Driver) updateClusterAndWait(ctx context.Context, state state) error {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(state.Region),
+		Credentials: credentials.NewStaticCredentials(
+			state.ClientID,
+			state.ClientSecret,
+			state.SessionToken,
+		),
+	})
+	if err != nil {
+		return err
+	}
+	svc := eks.New(sess)
+	output, err := svc.UpdateClusterVersionWithContext(ctx, &eks.UpdateClusterVersionInput{
+		Name:    aws.String(state.DisplayName),
+		Version: aws.String(state.KubernetesVersion),
+	})
+	if err != nil {
+		return err
+	}
+
+	return d.waitForClusterUpdateReady(ctx, svc, state, *output.Update.Id)
+}
+
+func (d *Driver) waitForClusterUpdateReady(ctx context.Context, svc *eks.EKS, state state, updateID string) error {
+	logrus.Infof("waiting for update id[%s] state", updateID)
+	var update *eks.DescribeUpdateOutput
+	var err error
+
+	status := ""
+	for status != "Successful" {
+		time.Sleep(30 * time.Second)
+
+		logrus.Infof("Waiting for cluster update to finish updating")
+
+		update, err = svc.DescribeUpdateWithContext(ctx, &eks.DescribeUpdateInput{
+			Name:     aws.String(state.DisplayName),
+			UpdateId: aws.String(updateID),
+		})
+		if err != nil {
+			return fmt.Errorf("error polling cluster update state: %v", err)
+		}
+
+		if update.Update == nil {
+			return fmt.Errorf("no cluster update status was returned")
+		}
+
+		status = *update.Update.Status
+	}
+
+	return nil
+}
+
+func getAMIs(ctx context.Context, ec2svc *ec2.EC2, state state) string {
```

Comment:
2 issues with this function:
1) I don't think this is working.  I tried running this locally (us-west-2) and it's getting back 0 image results.  I believe that this is because the query is looking for a prefix + k8s version, `amazon-eks-node-1.10`, but the images are actually prefix + k8s version + internal amazon version `amazon-eks-node-1.10-v20190109`.  We need to fix this query.
2) I think we should default to using our internal hardcoded AMIs, then, if a region is not found, query for one to use. My reasoning for this is that it is possible that amazon will introduce breaking changes into AMIs in the future.  We can protect ourselves against this possibility by only using AMIs that are "known good" at the time of writing the code.  We will still fall back to a dynamic look up, but only for regions that there is no known AMI for.

Reply:
Addressed